gui: Mark devices that haven't connected for a long time (fixes #7703)

tomasz1986 opened the pull request:
Currently, a disconnected device is marked as "Disconnected" without any
consideration whether the state is just temporary or rather it has been
like that for a long time, potentially requiring user intervention.

This commit adds a new state called "Disconnected (Inactive)" which is
shown for devices that have been disconnected for longer than a week. It
also changes the "Last seen" information, so that "Never" is used only
when the device hasn't connected at all, and the exact date is displayed
otherwise.

Additionally, when the date is older than 1 week, a note about that is
displayed below the date. Furthermore, when the date becomes older than
1 month, the note text colour changes to orange, and when it exceeds 1
year, the colour changes again to red. This is done to provide the user
with a visual clue that something may potentially be wrong and requires
a manual fix.

Signed-off-by: Tomasz Wilczyński <twilczynski@naver.com>

### Screenshots

#### Before

![image](https://user-images.githubusercontent.com/5626656/191123553-536af9b7-aafd-4274-9bc9-5b23c9051503.png)

#### After (ignore the date)

![image](https://user-images.githubusercontent.com/5626656/191123454-5f8c9113-3fc4-4723-a319-b43393a56cfb.png)

![image](https://user-images.githubusercontent.com/5626656/191123465-170117b1-c2d2-4693-852b-abafe1042412.png)

![image](https://user-images.githubusercontent.com/5626656/191123471-cdc86a5a-0850-4d7e-8e57-564789bd6ac6.png)

calmh reviewed (dismissed):
Looks reasonable to me! (Giving others a chance to weigh in, if nobody does feel free to merge.)

tomasz1986 commented:
One thing that I am worried about is that "stale" may be tricky to translate.

calmh commented:
Indeed. In Swedish it essentially becomes either just "old" (no real value judgement) or something akin to "mouldy"...

tomasz1986 commented:
I usually like to check for technical terms in https://microsoft.com/language, which is obviously a little biased and limited to Windows and other Microsoft products, but the translations there are generally done by professional translators and hence good quality. In case of stale though, the Polish translation is more like "outdated", "inactive", or "obsolete", which isn't exactly the same.

Although, to tell the truth, "inactive" doesn't sound bad, does it? Is it better than "stale"? Definitely easier to understand and translate, that's for sure.

On the other hand, the generic dictionary translation is similar to yours, and it's more about food (= not fresh) or something not interesting (= boring).

acolomb commented:
Agree that stale is difficult to translate and not very accurate. How about something involving "gone", "long gone" or "missing"? The latter is a common military term (MIA, missing in action) in English, but probably just as hard to translate or grasp. But maybe an inspiration?

imsodin commented:
What about "Long disconnected" or "Disconnected for a long time" (latter seems too long, former might not really be English?). I wouldn't add more information to the "title bar" - I think that should stay as minimal as possible. On that note I also dislike the "more than a" line - to me that uses space for very little added value (the date already conveys the info). If you want the color thing, that could go on the date and that "more than" thing into a tooltip. However just a weak personal oppinion and suggestion, feel free to disregard. As in only engage the bikeshedding if you want, I am totally fine with merging as is.

calmh commented:
`Gone (long time no see)`

acolomb commented:
Good direction, @calmh ... I guess we don't need the term "Disconnected" at all when "Gone" already implies that.

tomasz1986 commented:
Thank you very much for all the ideas :grin:! I will experiment and see what fits best here. I agree that the "Disconnected (something)" label should be rather short though, as very long strings are already problematic (e.g. they can overflow breaking the GUI, or they become truncated hiding useful information).

> Another thought: why not just put a time directly in those parens? "Disconnected (8 days ago)" where anything below one week is not shown.

I like this suggestion quite a bit, but it may be impossible to properly translate as well, as in some languages (e.g. Polish) the noun form changes depending on the number 😕 (and here we would end up with a lot of different forms, i.e. "day(s)", "week(s)", "month(s)", "year(s)", etc.). Also, it could potentially be too long to display in full.

> On that note I also dislike the "more than a" line - to me that uses space for very little added value (the date already conveys the info). If you want the color thing, that could go on the date and that "more than" thing into a tooltip.

I actually tested doing just that first, but the coloured date with nothing else looked kind of poor in terms of aesthetics :disappointed:. That's why I switched to the current format instead. Obviously, the note below the main part is based on the lines "Reduced by ignore patterns" and "Altered by ignore deletes" that are used similarly under folder local state.

acolomb commented on gui/default/index.html:
```diff
@@ -744,7 +744,9 @@ <h4 class="panel-title">
                     <span ng-switch-when="disconnected"><span class="hidden-xs" translate>Disconnected</span><span class="visible-xs" aria-label="{{'Disconnected' | translate}}"><i class="fas fa-fw fa-power-off"></i></span></span>
+                    <span ng-switch-when="disconnected-inactive"><span class="hidden-xs" translate>Disconnected (Inactive)</span><span class="visible-xs" aria-label="{{'Disconnected (Inactive)' | translate}}"><i class="fas fa-fw fa-power-off"></i></span></span>
                     <span ng-switch-when="unused-disconnected"><span class="hidden-xs" translate>Disconnected (Unused)</span><span class="visible-xs" aria-label="{{'Disconnected (Unused)' | translate}}"><i class="fas fa-fw fa-unlink"></i></span></span>
+                    <span ng-switch-when="unused-disconnected-inactive"><span class="hidden-xs" translate>Disconnected (Unused, Inactive)</span><span class="visible-xs" aria-label="{{'Disconnected (Unused, Inactive)' | translate}}"><i class="fas fa-fw fa-unlink"></i></span></span>
```
This one I find a bit confusing. If we don't share anything with a device, does it matter whether the connection had been established short or long ago?

```suggestion
                    <span ng-switch-when="unused-disconnected-inactive"><span class="hidden-xs" translate>Disconnected (Unused)</span><span class="visible-xs" aria-label="{{'Disconnected (Unused)' | translate}}"><i class="fas fa-fw fa-unlink"></i></span></span>
```

tomasz1986 replied:
Done.

acolomb replied:
Your latest commit removed the whole case though, so nothing will be displayed for `unused-disconnected-inactive`.  We should just display the same label as for `unused-disconnected`, since both are still true.

tomasz1986 replied:
Thanks for noticing! I've fixed this now by reverting to what the code looked originally. What do you think?

acolomb replied:
I meant dropping the `, Inactive` part within parentheses. Just use the same label `Disconnect (Unused)` as stated above.

tomasz1986 replied:
Done. I've reverted the JS change and re-added the HTML. There's a lot of duplicated code in that whole HTML block, which I think could be simplified quite a bit, but I'll leave this for a different PR.

tomasz1986 commented:
I've changed the name from "stale" to "inactive". I'm not 100% sure if it's the best option, but at least it seems similar enough in meaning, and it should be much easier to translate too. I did try the other alternatives out and browsed through various English synonym dictionaries looking for other examples, but I really think this should only be a single word (like the existing "unused"), so the number of possibilities is limited here.

I've also added an exception not to calculate the date when it equals the Unix epoch (i.e. 1970-01-01), and I've added yet another case of "Disconnected (Unused, Inactive)", which was missing previously.

acolomb reviewed (commented):
Haven't looked at the result in action, just commenting code. I think it's going in the right direction.

imsodin commented on gui/default/syncthing/core/syncthingController.js:
```diff
@@ -1094,7 +1096,11 @@ angular.module('syncthing.core')
             }
 
             // Disconnected
-            return status + 'disconnected';
+            if ($scope.deviceStats[deviceCfg.deviceID].lastSeenDays >= 7) {
+                return status + 'disconnected-inactive';
+            } else {
+                return status + 'disconnected';
+            }
```
Just a little nit: I'd check for unused here and not add the `-inactive` suffix then. The two states are the exact same in the web UI and I agree with that semantically: I think unused is the stronger/more relevant state than inactive, and having both doesn't add much IMO.

tomasz1986 replied:
Done.

tomasz1986 commented on gui/default/syncthing/core/syncthingController.js:
```diff
@@ -1096,7 +1096,7 @@ angular.module('syncthing.core')
 
             // Disconnected
-            if ($scope.deviceStats[deviceCfg.deviceID].lastSeenDays >= 7) {
+            if (status !== 'unused-' && $scope.deviceStats[deviceCfg.deviceID].lastSeenDays >= 7) {
```
Done.

imsodin commented on gui/default/syncthing/core/syncthingController.js:
```diff
@@ -1072,8 +1074,11 @@ angular.module('syncthing.core')
 
         $scope.deviceStatus = function (deviceCfg) {
             var status = '';
-
             if ($scope.deviceFolders(deviceCfg).length === 0) {
+                var unused = true;
+            }
+
+            if (unused) {
```
That might be correct JS, but scoping at least looks wrong - something like this I'd expect

```
            var unused =($scope.deviceFolders(deviceCfg).length === 0

            if (unused) {
```

tomasz1986 replied:
Done. Is there any difference in scope though? Variables declared with `var` are available inside the whole function anyway.

imsodin replied:
That's what I meant: It likely is correct and doesn't matter, but looks wrong and definitely isn't elegant (to exagerate you never want to do `if condition { conditionVariable = true } else { conditionVariable = false}` when you can just do `conditionVariable = condition`.

rubin55 commented:
I just noticed this change after updating today. One of my hosts has been off for about 6 months. I suppose a lot of users will see this for the first time: Question: what are the consequences of a host being unavailable for a longer time? Wouldn't it just catch up with a lot of sync traffic? Is there a risk of dataloss?

tomasz1986 commented:
The purpose of the message is just to provide information. There are valid use cases for having devices not connected for a longer period of time. However, when there are many devices, it's also easy to miss a specific one that haven't been syncing even though it was supposed to, hence the new "inactive" state that is visible in the main part of the Web GUI.

There's no risk of "data loss" per se, but if the device was modifying files that were supposed to sync right away yet for some reason they haven't, you may then end up with a lot of conflicts once it finally connects.

When it comes to the colour, it will change to red as soon as the time period reaches 365 days :wink:.
